text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+## Breaking changes in CX Protocol
+
+Probably the most prominent scenario for causing breakage is pushing an incompatible change to the cx-protocol itself.
+This can, for example, be the removal of a property.
+
+> See concrete [example](#rename-target-property-in-containerImageAssetMetadataEntry).
+
+The proposed solution for this case is to run API compatibility checks using `jsii-diff`.
+We would treat `cx-api` as a regular `jsii` module that needs to maintain compatibility to its consumers, which is the CLI.
+
+This will make sure that when the CLI pulls in new versions of `cx-api` (on upgrades), it will properly consume older cloud assemblies.
+
+## Breaking changes in CLI
+
+Another way to break CLI upgrades, is to simply change the CLI in a breaking way. This can, for example, be a change to the docker build command.
+
+> See concrete [example](#remove---target-from-docker-build-command).
+
+The proposed solution for this case is to run standard regression tests. That is, run old CLI integration tests on the latest code (Framework and CLI).
+
+This will make sure that old CLI functionality is still working, assuming of course it was covered by our integration tests.
+
+# Detailed Design
+
+There are 2 mechanisms we need to implement:
+
+## API Compatibility Checks
+
+We need to make sure the objects that make up what we refer to as the `cx-protocol`, do not change in a breaking way.
+We should be able to leverage `jsii` to accomplish this. For example, this is an excerpt from the `.jsii` spec of the `cx-api` module:
+
+```json
+"name": "ContainerImageAssetMetadataEntry",
+"properties": [
+    {
+        "abstract": true,
+        "docs": {
+            "stability": "experimental",
+            "summary": "Path on disk to the asset."
+        },
+        "immutable": true,
+        "locationInModule": {
+            "filename": "lib/assets.ts",
+            "line": 48
+        },
+        "name": "path",
+        "type": {
+            "primitive": "string"
+        }
+    }
+]
+```
+
+This means that, if we remove the `path` property from `ContainerImageAssetMetadataEntry`, jsii should produce an error, and indeed:
+
+```console
+❯ yarn compat
+PROP @aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path: has been removed [removed:@aws-cdk/cx-api.ContainerImageAssetMetadataEntry.path]
+```
+
+Currently, the `cx-api` module contains a lot of stuff that aren't strictly related to the protocol.
+
+> For example, `AssemblyBuildOptions`, which is used during synthesis by the framework. It is not used by the CLI (nor it should).
+
+This creates clutter, and might also cause issues with applying `jsii` compatibility checks.
+
+In addition, according to our new compatibility model, we need to make sure the user gets the correct error
+when the CLI version is smaller than the `Cloud-Assembly` version.
+
+The plan therefore is as follows:
+
+### Step 1: Create a separate package called `cx-protocol`
```

Comment:
Let's call it `@aws-cdk/cloud-assembly` so we are very explicit about what exactly is defined in this API.

Reply:
decide to go with `cloud-assembly-schema`